### PR TITLE
JW Player

### PR DESCRIPTION
**Pull request by anoblet**
https://app.clickup.com/t/3tnvw4y

### CODE REVIEW

**Review by pawelkmpt — COMMENTED** [review]
You squashed commits, nice 👍 

Please polish one thing in commit messages:

```diff
- [cxl-jw-player]
+ cxl-jw-player
```

**Comment by pawelkmpt** [comment]
I managed to run a player on WPS. It doesn't show transcript though, but I didn't dig into details yet as I'm coming with security thought.

Component requires to pass `api-secret` which makes it a public value and it should not happen. Otherwise everyone can use it. We should rather sign and obtain necessary data on the backend side.

@lkraav @anoblet your thoughts?

File: packages/cxl-ui/src/components/cxl-jw-player/mixins/BaseMixin.js [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+    }
+
+    get _scriptUrl() {
+      if (!this.libraryId && !this.librarySource) return false;
+
+      let scriptUrl;
+
+      if (this.libraryId) {
+        if (this.isPublic) {
+          scriptUrl = `https://content.jwplatform.com/libraries/${this.libraryId}.js`;
```
Comment (pawelkmpt) [inline packages/cxl-ui/src/components/cxl-jw-player/mixins/BaseMixin.js:73]: Other places use `https://cdn.jwplayer.com` URL, shouldn't we use it here too?

File: packages/cxl-ui/src/components/cxl-jw-player/mixins/BaseMixin.js [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+    }
+
+    async _getMedia() {
+      if (!this.mediaId && !this.mediaSource) return false;
+
+      let response;
+
+      if (this.mediaId) {
+        if (this.isPublic) {
+          response = await fetch(`https://cdn.jwplayer.com/v2/media/${this.mediaId}`);
```
Comment (pawelkmpt) [inline packages/cxl-ui/src/components/cxl-jw-player/mixins/BaseMixin.js:116]: `https://cdn.jwplayer.com` is repeated in the code multiple times. It should be defined as constant in the one place and used across the component

**Review by pawelkmpt — CHANGES_REQUESTED** [review]

File: packages/cxl-ui/src/components/cxl-jw-player/mixins/TranscriptMixin.js [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+import { parseSync } from 'subtitle';
+
+export function TranscriptMixin(BaseClass) {
+  class Mixin extends BaseClass {
+    _debouncedSearch;
+
+    _mark;
+
+    @property({ reflect: true, type: Boolean }) captions = false;
+    @property({ attribute: 'has-captions', reflect: true, type: Boolean }) hasCaptions = false;
```
Comment (pawelkmpt) [inline packages/cxl-ui/src/components/cxl-jw-player/mixins/TranscriptMixin.js:13]: Captions can be enabled in video settings:
Screenshot 2024-09-20 at 14 10 32

but we're in the "transcript" class so, shouldn't it be named "has-transcript"?

File: packages/cxl-ui/src/components/cxl-jw-player/mixins/index.js [hunk 4]
```diff
@@ -0,0 +1,5 @@
+export { BaseMixin } from './BaseMixin';
+export { ChapterNavigationMixin } from './chapter-navigation/index.js';
+export { NextUpMixin } from './NextUpMixin';
+export { StateMixin } from './StateMixin';
+export { TranscriptMixin } from './TranscriptMixin';
```
Comment (lkraav) [inline packages/cxl-ui/src/components/cxl-jw-player/mixins/index.js:5]: "Kebab-case everywhere" possible?

Maybe we can also install https://github.com/loeffel-io/ls-lint one day.

File: packages/cxl-ui/src/components/cxl-jw-player/mixins/TranscriptMixin.js [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+        this._jwPlayer.addButton(
+          `<svg class="jw-player-button" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" preserveAspectRatio="xMidYMid meet" aria-hidden="true" viewBox="0 0 1000 1000"><g><path d="M662 603l131 131c16 16 16 42 0 59-16 16-43 16-59 0l-131-131C562 691 512 708 458 708c-138 0-250-112-250-250 0-138 112-250 250-250 138 0 250 112 250 250 0 54-17 104-46 145zM458 646c104 0 188-84 188-188S562 271 458 271 271 355 271 458s84 188 187 188z"></path></g></svg>`,
+          'Transcript',
+          this._toggleTranscript.bind(this),
+          'toggle-transcript'
+        );
+      } else {
+        this.captions = false;
+        this._jwPlayer.removeButton('toggle-transcript');
+      }
```
Comment (pawelkmpt) [inline packages/cxl-ui/src/components/cxl-jw-player/mixins/TranscriptMixin.js:160]: I don't see this button anywhere :(

**Comment by pawelkmpt** [comment]
> Component requires to pass `api-secret` which makes it a public value and it should not happen. Otherwise everyone can use it. We should rather sign and obtain necessary data on the backend side.
> 
> @lkraav @anoblet your thoughts?
> 
> EDIT We have `/wp-json/cxl/jwp/content-token` REST API endpoint for signing resource ID. It was created for JW Player app. Maybe can be used here too?

Maybe component should get signed URL as param, e.g. https://cdn.jwplayer.com/v2/media/fVRbGa56?sources=hls&token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJleHAiOjE3MjcwOTk2MTQsInJlc291cmNlIjoidjIvbWVkaWEvZlZSYkdhNTYifQ.ZWjguN1Rya0rHXz-kH3TomHINFwp_-NA6oBcnDwrNAU and inside fetch data, process it and display what needs to be displayed?

Using REST API endpoint requires authentication too so some kind of secret would be still needed. With signing on the server side and providing signed URL we have this problem solved.

File: packages/cxl-ui/src/components/cxl-jw-player/mixins/BaseMixin.js (hunk 1 above)
Comment (anoblet) [reply packages/cxl-ui/src/components/cxl-jw-player/mixins/BaseMixin.js:73]: I can refactor this.

**Comment by lkraav** [comment]
> Maybe

Undoubtedly there's documented best practices out there for this age old problem, I'd welcome a link to a worthy guide?

File: packages/cxl-ui/src/components/cxl-jw-player/mixins/TranscriptMixin.js (hunk 3 above)
Comment (anoblet) [reply packages/cxl-ui/src/components/cxl-jw-player/mixins/TranscriptMixin.js:13]: Captions and transcript are pretty much the same thing. There was a requirement for the full text to be available to read. I prefer naming everything captions right now, I believe Syed renamed a few things and didn't finish it.

File: packages/cxl-ui/src/components/cxl-jw-player/mixins/index.js (hunk 4 above)
Comment (anoblet) [reply packages/cxl-ui/src/components/cxl-jw-player/mixins/index.js:5]: I'm trying to remember the reason for using pascal case vs. kebab case for mixins. I'll take a look and find a standard.

File: packages/cxl-ui/src/components/cxl-jw-player/mixins/TranscriptMixin.js (hunk 5 above)
Comment (anoblet) [reply packages/cxl-ui/src/components/cxl-jw-player/mixins/TranscriptMixin.js:160]: It's only available for videos with captions. It's the magnifying glass at the bottom right hand corner of the video.

deploy-preview-433--conversionxl-aybolit netlify app__path=_story_cxl-ui-cxl-jw-player--default

**Comment by anoblet** [comment]
>Component requires to pass api-secret which makes it a public value and it should not happen. Otherwise everyone can use it. We should rather sign and obtain necessary data on the backend side.

We saw this as a potential issue. Instead of providing `api-secret`, along with `library-id`, and `media-id`. You can omit `api-secret`, and use `library-source`, and `media-source` instead where the JWT signing happens on the backend, and the full URL is provided to the component.

**Comment by pawelkmpt** [comment]
> Undoubtedly there's documented best practices out there for this age old problem, I'd welcome a link to a worthy guide?

ChatGPT suggests using signing API: https://chatgpt.com/share/66f18132-43dc-8002-95ab-4f699f384976

As can seen above, I considered it too but managing REST API endpoint and authenticating request seems to be overengineering in our case when I can just output HTML with webcomponent having already signed URL. And cached URL

**Comment by anoblet** [comment]
>As API request or straighforward <cxl-jw-player url="<signed_url>">?

This would be the full signed URL

**Comment by pawelkmpt** [comment]
> > As API request or straighforward ?
> 
> This would be the full signed URL. There are two different types of signed URL, JWT and non JWT, we would need them both: https://docs.jwplayer.com/platform/reference/protect-your-content-with-signed-urls

Oh, indeed. Good catch @anoblet. I didn't see it before but now see JWT is used for a video and non JWP for library. So we'd need two arguments:
* `video-url` or `content-url`
* `lib-url`

or it's more complex?

Anyway, please adjust the component for the new approach. Save in new commits, do not overwrite existing just in case we'd like to get back to the past version @anoblet

**Comment by anoblet** [comment]
This is already supported with `library-source`, `media-source`, and `playlist-source` attributes. Sample code: https://github.com/conversionxl/aybolit/blob/2a873b429ed5c27c6c4145af8499466e76cda785/packages/cxl-ui/src/components/cxl-jw-player/mixins/BaseMixin.js#L111

File: packages/cxl-ui/src/components/cxl-jw-player/index.html.js [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+                    <vaadin-text-field
+                      @input=${this._debouncedSearch}
+                      clear-button-visible
+                      id="search"
+                      placeholder="Type and press enter to search"
+                      theme="small"
+                    >
+                      <vaadin-icon slot="prefix" icon="lumo:search"></vaadin-icon>
+                    </vaadin-text-field>
+                    <div id="searchResultCount">
```
Comment (pawelkmpt) [inline packages/cxl-ui/src/components/cxl-jw-player/index.html.js:24]: We don't use `camelCase` for HTML classes and IDs

```suggestion
                    <div id="search-result-count">
```

**Review by pawelkmpt — CHANGES_REQUESTED** [review]
2 more things on ClickUp

File: packages/cxl-ui/src/components/cxl-jw-player/index.html.js [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+                      ${this._isSearchMinimumLength
+                        ? html`<span>(${this._matches} matches)</span>`
+                        : nothing}
+                    </div>
+                    <vaadin-checkbox
+                      @change=${this._toggleShouldScroll}
+                      ?checked=${this._shouldScroll}
+                      label="Scroll"
+                    ></vaadin-checkbox>
```
Comment (pawelkmpt) [inline packages/cxl-ui/src/components/cxl-jw-player/index.html.js:33]: It's not visible. Is it needed?

**Comment by pawelkmpt** [comment]
@anoblet I have a question about using `has-captions` attribute. When I add it to the component, transcript disappears. What is the correct usage?

Screenshot 2024-10-01 at 12 47 08

**Comment by pawelkmpt** [comment]
@anoblet I have a suspicion `CXLJWPlayerElement` is compiled into `cxl-ui-institue.js` package. We also have separate `cxl-ui-jwplayer.js` and component should be imported only in the latter one.

File: packages/cxl-ui/src/components/cxl-jw-player/index.html.js (hunk 6 above)
Comment (anoblet) [reply packages/cxl-ui/src/components/cxl-jw-player/index.html.js]: This should be fixed.

File: packages/cxl-ui/src/components/cxl-jw-player/mixins/TranscriptMixin.js [hunk 8]
```diff
@@ -131,6 +140,20 @@ export function TranscriptMixin(BaseClass) {
       this._setupTranscript();
 
       this._jwPlayer.on('playlistItem', this._setupTranscript.bind(this));
+
+      this.shadowRoot.querySelector('vaadin-text-field').addEventListener('keyup', (e) => {
+        if(e.key === 'Enter') {
```
Comment (pawelkmpt) [inline packages/cxl-ui/src/components/cxl-jw-player/mixins/TranscriptMixin.js:145]: Use yoda condition

File: packages/cxl-ui/src/components/cxl-jw-player/mixins/TranscriptMixin.js [hunk 9]
```diff
@@ -131,6 +140,20 @@ export function TranscriptMixin(BaseClass) {
...
+          if(this._searchResults.length) {
+            if(this._searchIndex === this._searchResults.length - 1) {
+              this._searchIndex = 0;
+            }
```
Comment (pawelkmpt) [inline packages/cxl-ui/src/components/cxl-jw-player/mixins/TranscriptMixin.js:149]: Correct formatting

```diff
- if(
+ if (
```

**Review by pawelkmpt — APPROVED** [review]

**Pull request merged**